File: ios/PretextConstants.swift
import Foundation

internal let newlineToken = "\n"
internal let whiteSpaceNormal = "normal"
internal let whiteSpacePre = "pre"
internal let wordBreakNormal = "normal"
internal let wordBreakBreakAll = "break-all"
internal let breakBehaviorNever = "never"
internal let inlineSegmentKindBox = "box"
internal let layoutEngineIosCoreText = "ios_core_text"
internal let layoutEngineIosManualTokenFallback = "ios_manual_token_fallback"
internal let fallbackReasonManualHeightEstimate = "manual_height_estimate"
internal let heightMetricSourcePlatformTextEngineMetrics = "platform_text_engine_metrics"
internal let ruleLayerPretextNative = "pretext_native_rules"
internal let breakKindHard = "hard_break"
internal let breakKindNativeSoft = "native_soft_break"
internal let breakSourceNewline = "source_newline"
internal let driftAlgorithmRule = "algorithm_rule_drift"
internal let driftClusterBoundary = "cluster_boundary_drift"
internal let driftEmojiMetric = "emoji_metric_drift"
internal let driftEngine = "engine_drift"
internal let driftFallbackFont = "fallback_font_drift"
internal let driftHeightMetric = "height_metric_drift"
internal let driftLineBreakStrategy = "line_break_strategy_drift"
internal let driftLocaleMetric = "locale_metric_drift"
internal let driftPadding = "padding_drift"
internal let heightMetricDrivers = [
    "font_metrics",
    "explicit_line_height",
    "fallback_font",
    "emoji_fallback",
    "locale",
    "include_font_padding",
    "line_break_strategy",
]
internal let zeroWidthJoiner: UInt32 = 0x200D

internal func isCombiningMark(_ scalar: UnicodeScalar) -> Bool {
    CharacterSet.nonBaseCharacters.contains(scalar)
}

internal func isVariationSelector(_ scalar: UnicodeScalar) -> Bool {
    (0xFE00...0xFE0F).contains(scalar.value) || (0xE0100...0xE01EF).contains(scalar.value)
}

internal func isEmojiModifier(_ scalar: UnicodeScalar) -> Bool {
    (0x1F3FB...0x1F3FF).contains(scalar.value)
}

internal func isRegionalIndicator(_ scalar: UnicodeScalar) -> Bool {
    (0x1F1E6...0x1F1FF).contains(scalar.value)
}

internal func isIndicVirama(_ scalar: UnicodeScalar) -> Bool {
    [
        0x094D,
        0x09CD,
        0x0A4D,
        0x0ACD,
        0x0B4D,
        0x0BCD,
        0x0C4D,
        0x0CCD,
        0x0D4D,
        0x0DCA,
        0x0E3A,
        0x0F84,
        0x1039,
        0x103A,
        0x1714,
        0x1734,
        0x17D2,
        0x1A60,
    ].contains(scalar.value)
}

internal func isRtlScalar(_ scalar: UnicodeScalar) -> Bool {
    if CharacterSet.decimalDigits.contains(scalar) || isCombiningMark(scalar) {
        return false
    }

    return (0x05D0...0x05EA).contains(scalar.value)
        || (0x0620...0x064A).contains(scalar.value)
        || (0x066E...0x066F).contains(scalar.value)
        || (0x0671...0x06D3).contains(scalar.value)
        || scalar.value == 0x06D5
        || (0x06E5...0x06E6).contains(scalar.value)
        || (0x06EE...0x06EF).contains(scalar.value)
        || (0x06FA...0x06FC).contains(scalar.value)
        || scalar.value == 0x06FF
        || (0x0700...0x074F).contains(scalar.value)
        || (0x0750...0x077F).contains(scalar.value)
        || (0x0780...0x07BF).contains(scalar.value)
        || (0x07C0...0x07FF).contains(scalar.value)
        || (0x0800...0x083F).contains(scalar.value)
        || (0x0840...0x085F).contains(scalar.value)
        || (0x0860...0x086F).contains(scalar.value)
        || (0x0870...0x089F).contains(scalar.value)
        || (0x08A0...0x08FF).contains(scalar.value)
        || (0x10AC0...0x10AFF).contains(scalar.value)
        || (0x10D00...0x10D3F).contains(scalar.value)
        || (0xFB1D...0xFDFF).contains(scalar.value)
        || (0x1E800...0x1E8DF).contains(scalar.value)
        || (0x1E900...0x1E95F).contains(scalar.value)
        || (0xFE70...0xFEFF).contains(scalar.value)
}

internal func isLtrScalar(_ scalar: UnicodeScalar) -> Bool {
    CharacterSet.letters.contains(scalar) && !isRtlScalar(scalar)
}

File: android/src/main/java/com/margelo/nitro/pretext/PretextConstants.kt
package com.margelo.nitro.pretext

internal const val NEWLINE_TOKEN = "\n"
internal const val WHITE_SPACE_NORMAL = "normal"
internal const val WHITE_SPACE_PRE = "pre"
internal const val WORD_BREAK_NORMAL = "normal"
internal const val WORD_BREAK_BREAK_ALL = "break-all"
internal const val BREAK_BEHAVIOR_NEVER = "never"
internal const val INLINE_SEGMENT_KIND_BOX = "box"
internal const val LAYOUT_ENGINE_ANDROID_MEASURED_TEXT_LINE_BREAKER =
  "android_measured_text_line_breaker"
internal const val LAYOUT_ENGINE_ANDROID_STATIC_LAYOUT_COMPAT = "android_static_layout_compat"
internal const val LAYOUT_ENGINE_ANDROID_LEGACY_FALLBACK = "android_legacy_fallback"
internal const val FALLBACK_REASON_MANUAL_HEIGHT_ESTIMATE = "manual_height_estimate"
internal const val FALLBACK_REASON_STATIC_LAYOUT_COMPAT = "static_layout_compat"
internal const val HEIGHT_METRIC_SOURCE_PLATFORM_TEXT_ENGINE_METRICS =
  "platform_text_engine_metrics"
internal const val RULE_LAYER_PRETEXT_NATIVE = "pretext_native_rules"
internal const val BREAK_KIND_HARD = "hard_break"
internal const val BREAK_KIND_NATIVE_SOFT = "native_soft_break"
internal const val DRIFT_CLUSTER_BOUNDARY = "cluster_boundary_drift"
internal const val DRIFT_ALGORITHM_RULE = "algorithm_rule_drift"
internal const val DRIFT_EMOJI_METRIC = "emoji_metric_drift"
internal const val DRIFT_ENGINE = "engine_drift"
internal const val DRIFT_FALLBACK_FONT = "fallback_font_drift"
internal const val DRIFT_HEIGHT_METRIC = "height_metric_drift"
internal const val DRIFT_LINE_BREAK_STRATEGY = "line_break_strategy_drift"
internal const val DRIFT_LOCALE_METRIC = "locale_metric_drift"
internal const val DRIFT_PADDING = "padding_drift"
internal const val ZERO_WIDTH_JOINER = 0x200D

internal val HEIGHT_METRIC_DRIVERS =
  arrayOf(
    "font_metrics",
    "explicit_line_height",
    "fallback_font",
    "emoji_fallback",
    "locale",
    "include_font_padding",
    "line_break_strategy",
  )

internal fun isCombiningMark(codePoint: Int): Boolean {
  return when (Character.getType(codePoint)) {
    Character.NON_SPACING_MARK.toInt(),
    Character.COMBINING_SPACING_MARK.toInt(),
    Character.ENCLOSING_MARK.toInt() -> true
    else -> false
  }
}

internal fun isVariationSelector(codePoint: Int): Boolean {
  return codePoint in 0xFE00..0xFE0F || codePoint in 0xE0100..0xE01EF
}

internal fun isEmojiModifier(codePoint: Int): Boolean {
  return codePoint in 0x1F3FB..0x1F3FF
}

internal fun isRegionalIndicator(codePoint: Int): Boolean {
  return codePoint in 0x1F1E6..0x1F1FF
}

internal fun isIndicVirama(codePoint: Int): Boolean {
  return codePoint in setOf(
    0x094D,
    0x09CD,
    0x0A4D,
    0x0ACD,
    0x0B4D,
    0x0BCD,
    0x0C4D,
    0x0CCD,
    0x0D4D,
    0x0DCA,
    0x0E3A,
    0x0F84,
    0x1039,
    0x103A,
    0x1714,
    0x1734,
    0x17D2,
    0x1A60,
  )
}
